### PR TITLE
Use coroutine in RecurringTask

### PR DESCRIPTION
**Pull request by cooltey**
This PR also updates the usage of `WikipediaApp.instance` and moves it to the `RecurringTask` class.

### CODE REVIEW

File: app/src/main/java/org/wikipedia/activity/BaseActivity.kt [hunk 1]
```diff
@@ -97,7 +97,7 @@ abstract class BaseActivity : AppCompatActivity(), ConnectionStateMonitor.Callba
         }
 
         // Conditionally execute all recurring tasks
-        RecurringTasksExecutor(WikipediaApp.instance).run()
+        RecurringTasksExecutor(lifecycleScope).run()
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/activity/BaseActivity.kt:100]: Actually I don't think the Recurring Tasks should be bound to the `lifecycleScope` of this activity (they should be allowed to run independently of any activity). Can we make them use `MainScope` instead?

File: app/src/main/java/org/wikipedia/recurring/DailyEventTask.kt [hunk 2]
```diff
@@ -1,22 +1,21 @@
...
-class DailyEventTask(context: Context) : RecurringTask() {
-    override val name = context.getString(R.string.preference_key_daily_event_time_task_name)
+class DailyEventTask(private val app: WikipediaApp) : RecurringTask() {
+    override val name = app.getString(R.string.preference_key_daily_event_time_task_name)
 
     override fun shouldRun(lastRun: Date): Boolean {
         return millisSinceLastRun(lastRun) > TimeUnit.DAYS.toMillis(1)
     }
 
     override fun run(lastRun: Date) {
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/recurring/DailyEventTask.kt:17]: The `run()` function and `runIfNecessary()` should probably be `suspend` functions (and run in `IO` dispatcher).

**Review by dbrant — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/wikipedia/recurring/DailyEventTask.kt (hunk 2 above)
Comment (cooltey) [reply app/src/main/java/org/wikipedia/recurring/DailyEventTask.kt:17]: Done!

File: app/src/main/java/org/wikipedia/activity/BaseActivity.kt (hunk 1 above)
Comment (cooltey) [reply app/src/main/java/org/wikipedia/activity/BaseActivity.kt:100]: I see. Done.

**Comment by dbrant** [comment]
👏

**Pull request merged**